articles/events_executor.md

```diff
@@ -0,0 +1,233 @@
+Let's consider a ROS 2 subscription as an example.
+The event will contain the `ExecutorEventType::SUBSCRIPTION_EVENT` label and an identifier to a `rclcpp::SubscriptionBase` object, and the executor will use them to get the `rclcpp::SubscriptionBase` object that needs to be passed to the `execute_subscription()` API.
+
+`Waitables` implement their own `execute()` API which can be called directly by the executor when a `ExecutorEventType::WAITABLE_EVENT` is received.
+
+The `EventsExecutor` should allow entities to push events into the queue while other events are being processed, i.e. without blocking the producers.
+The current implementation achieves that by having two separate `EventsQueue`: one for storage and accumulation of events and the other for their execution, and swapping the two whenever the executor wakes up.
+
+In order to implement the non-blocking variants of `spin()` it is possible to get from the `TimersManager` the duration until the first timer expires and then use this value as a timeout in the conditional wait for new events.
+
+The `EventsExecutor` should take care of correctly handling events that have been generated from entities that are not associated with the executor anymore, either because they have been removed or because they went out of scope.
+
+![Events Execution](../img/events_executor/execution.png)
+
+The `EventsExecutor` design allows to easily extend the execution of events from a single thread to a multi thread one, simply by creating new threads whenever an entity has to be executed.
+
+### Cleanup and ownership
+
+The `EventsExecutor` should not keep ownership of the nodes and of the entities that are associated to it.
+
+Whenever a node or a callback group are removed from the executor using the corresponding APIs (`remove_node()` and `remove_callback_group()`), the `EventsExecutor` needs to clean up the associated entities to make them stop sending events.
+Entities may be eventually added back again to the same executor or to a different one and they need to be able to start again producing events.
+This requires that the initialization procedure that is performed when entities are added to an executor is completely reversible.
+
+The events-based approach requires particular care in handling the case where entities are destroyed while still associated with an executor.
+An entity may push an event into the `EventsQueue` and then get immediately destroyed.
+This may happen before the `EventsExecutor` start to process those events or while it is processing them.
+
+The current implementation addresses this problem by providing an additional function pointer to each associated entity.
+This function pointer will be called whenever the entity's destructor is invoked and it will result in clearing the `EventsQueue` from any event that was generated by this entity. Note that this may block until the `EventsQueue` is done processing events.
```

Comment:
If the `EventsExecutor` does not own entities, what would prevent entity destruction during event processing (and the resulting deadlock)?

Reply:
The EventsExecutor does not own entities mainly to align with the current ROS 2 design paradigm where there are no APIs for removing/destroying entities.

There should not be a deadlock: the entity destructor and the event queue processing thread will all try to lock the same single mutex.

Assume the following: entity A and entity B are being destroyed and they both have events on the queue.
If A acquires the lock first, while holding this lock it can remove its own events from the queue.
Upon release the events executor may acquire the lock and it will execute B events.
Then B may acquire the lock and it will not have anything to remove.

There are definitely other solutions to handle this situation.

Reply:
In the current implementation, mutex prevents entity destruction during event processing:
https://github.com/irobot-ros/rclcpp/blob/irobot/add-events-executor/rclcpp/include/rclcpp/executors/events_executor.hpp#L216
https://github.com/irobot-ros/rclcpp/blob/irobot/add-events-executor/rclcpp/include/rclcpp/executors/events_executor.hpp#L225

Reply:
> In the current implementation, mutex prevents entity destruction during event processing

Hmm, that's true when destruction and execution happen in separate threads. But unless I'm missing something from your `rclcpp` based prototype (which I might), if any entity associated with the event executor is destroyed within one of the callbacks that consume events [while `execution_mutex_` is being held](https://github.com/irobot-ros/rclcpp/blob/6c55f0bd79ef806825f0edb1e4abd7402ece75ac/rclcpp/src/rclcpp/executors/events_executor.cpp#L68), on entity removal [it will attempt to lock the `execution_mutex_` again](https://github.com/irobot-ros/rclcpp/blob/6c55f0bd79ef806825f0edb1e4abd7402ece75ac/rclcpp/include/rclcpp/executors/events_executor.hpp#L225). And `execution_mutex_` isn't recursive.

Reply:
Yes, that situation can cause a deadlock and making that mutex recursive should be enough to handle it.

However, again I'm not sure what the use-case for that could be.
In the current design the event types are used only to  distinguish between entities.
This means that you would have a regular subscription that, depending on the content of the message it receives, could decide to destroy itself or maybe to destroy another entity?
I think that ROS does not really support this workflow as entities are usually handled with shared ownership.

Reply:
The `StaticSingleThreadedExecutor` has shared ownership of entities. This model does not allow a callback to destroy other entities, as the executor will always keep 1 reference to them. It actually does not really allow to destroy entities at all without first removing the node from the executor.
The `SingleThreadedExecutor` has weak ownership of entities. This model allows a callback to destroy other stuffs.

With the `EventsExecutor` we did something different from both.
The weak ownership model is not performant as the "fast path" require you to lock all the entities every time. I haven't measured this overhead, so maybe it's not really comparable to the other performance issues in ROS, but it's definitely something that eventually needs to be addressed.

I'm fully conscious that the ownership of entities is a weak point of the current implementation, but IMO is something not related to the `EventsExecutor` design in particular, but rather to its implementation or to the executors in general.
I would prefer a shared ownership model where additional APIs allow you to easily handle entities (e.g. add/remove from node and add/remove from executor).

Considering your scenario (a callback that destroys an entity) it could first remove that entity from the executor and then destroy it.
However this would require many more changes to ROS.

Reply:
tl;dr I agree with @hidmic here:

> Whether that's useful or desirable is of no consequence: the implementation must behave in a sane way.

To merge this, we should keep the existing ownership behavior, and therefore preserve the ability to "delete" (or start asynchronously deleting) entities that might be being used by the executor from within a callback.

Specifically, the executor should have weak ownership on the entities associated with it, and strong ownership when using them.

----

> I'm fully conscious that the ownership of entities is a weak point of the current implementation, but IMO is something not related to the EventsExecutor design in particular, but rather to its implementation or to the executors in general.

If you want it to be as much of a "drop-in replacement" as possible (and I think that's a good thing for comparing them and making it easier to integrate) then the `EventsExecutor` should follow the existing behavior of shared ownership (i.e. user owns, executor has weak ownership). If it really is as decoupled as you say, then changing the executor implementation AND the sharing behavior is changing two things at once, and is generally not desired.

> The weak ownership model is not performant as the "fast path" require you to lock all the entities every time. I haven't measured this overhead, so maybe it's not really comparable to the other performance issues in ROS, but it's definitely something that eventually needs to be addressed.

I think that should be evaluated and addressed separately from this proposed change in the executor design. It's not clear to me that "but it's definitely something that eventually needs to be addressed." is true.

> I would prefer a shared ownership model where additional APIs allow you to easily handle entities (e.g. add/remove from node and add/remove from executor).

I don't believe this is substantially different from the current approach, because it still doesn't let you remove items from within a callback. You'd really only be able to have `remove` (which doesn't work in a callback) and `async_remove`/`remove_later` which would work in a callback, but that's not any different than releasing ownership of the shared pointer within a callback, imo.

Reply:
> The `StaticSingleThreadedExecutor` has shared ownership of entities. This model does not allow a callback to destroy other entities, as the executor will always keep 1 reference to them. It actually does not really allow to destroy entities at all without first removing the node from the executor.

That's a good point. I see shared ownership of entities in [`rclcpp::experimental::ExecutableList`](https://github.com/ros2/rclcpp/blob/9c62c1c9463cd2accee57fe157125950df50f957/rclcpp/include/rclcpp/experimental/executable_list.hpp#L67-L86).

> Considering your scenario (a callback that destroys an entity) it could first remove that entity from the executor and then destroy it.
However this would require many more changes to ROS.

I agree with you that we're digging a bit too much into implementation details here, but how so? Again, I could be wrong, but considering that callback groups have weak ownership and that removing an entity from the `EventsExecutor` effectively drops the push event callback and prunes related events, you just have to ensure you synchronize properly (as mentioned above). It won't trigger an event, sure, but that's a non-issue. Explicitly removing a node is safe too since you use guard conditions. Your design is pretty solid in that sense.

As @wjwwood says, we can discuss ownership later on. I'd much rather get your work reviewed and merged instead of growing the scope right now.

Reply:
Ok, thank you all for the discussion.
We are updating the rclcpp implementation to have weak ownership of entities.
This will allow us also to measure the difference in performance between the two approaches.

Reply:
![image](https://user-images.githubusercontent.com/16389257/102402681-11917600-3fdd-11eb-8e16-9e6b2e6537eb.png)
These are the results including the implementation of the EventsExecutor having weak ownership of entities, and strong ownership when using them. We'll soon push this changes to the rclcpp PR.

Reply:
**Update:** We've improved the weak pointer implementation for the new ownership model. The performances are now much better, using just 5% more CPU than the no-ownership approach (which used raw pointers thus no `lock()` needed to get ownership) for the worst case scenario (our biggest topology at 100hz in a single core platform). The graph above has been update reflecting the latest performances.